### PR TITLE
fix(ir): make impure ibis.random() and ibis.uuid() functions return unique node instances

### PR DESCRIPTION
**Pull request by kszucs**
Resolves https://github.com/ibis-project/ibis/issues/8921

### CODE REVIEW

File: ibis/backends/bigquery/compiler.py [hunk 1]
```diff
@@ -698,3 +696,6 @@ def visit_CountDistinct(self, op, *, arg, where):
         if where is not None:
             arg = self.if_(where, arg, NULL)
         return self.f.count(sge.Distinct(expressions=[arg]))
+
+    def visit_RandomUUID(self, op, **kwargs):
```
Comment (cpcloud) [inline ibis/backends/bigquery/compiler.py:700]: Why do you need to split these out and handle `kwargs`? The existing `SIMPLE_OPS` loop should handle the case of unused arguments.

**Comment by cpcloud** [comment]
I think this is a bug fix, since it has to do with output correctness.

File: ibis/backends/bigquery/compiler.py (hunk 1 above)
Comment (kszucs) [reply ibis/backends/bigquery/compiler.py:700]: It actually renders the arguments.

File: ibis/expr/operations/generic.py [hunk 2]
```diff
@@ -183,6 +183,17 @@ class Constant(Scalar, Singleton):
...
+    uid: Optional[int] = None
+
+    def __init__(self, uid, **kwargs):
+        if uid is None:
+            uid = next(self._counter)
+        super().__init__(uid=uid, **kwargs)
+
+
 @public
 class TimestampNow(Constant):
```
Comment (cpcloud) [inline ibis/expr/operations/generic.py:198]: This also needs to subclass from `Impure` along with `DateNow`

**Review by cpcloud — CHANGES_REQUESTED** [review]

**Comment by kszucs** [comment]
> There's also `now` and `today` and probably others.

I am planning to address that in a follow-up.

**Comment by cpcloud** [comment]
Any reason why those can't be done in this PR?

**Comment by jcrist** [comment]
Are `now` and `today` impure in the same sense? Implemented using `CURRENT_TIMESTAMP` these will return the same value for every instance of the expression when the query executes (even though they may return a different value for two subsequent executions). In contrast `random()` and `uuid()` will generally return a new value for each call in an expression (some backends like clickhouse handle this a bit differently).

**Comment by cpcloud** [comment]
Ah, good point. I hadn't considered that `select now() = now()` whereas `select random() <> random()`.

Do we know whether that's true for all of our backends?

**Comment by cpcloud** [comment]
Anyway, that's sufficiently different that I think it can be done in a follow up.

**Comment by jcrist** [comment]
> Do we know whether that's true for all of our backends?

I wouldn't be surprised if it's broken for pandas/dask/polars, but semantically I think it _should_ be true for all backends.

File: ibis/backends/bigquery/compiler.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/bigquery/compiler.py:700]: Ah, right. Extra kwargs get rendered in the `SIMPLE_OPS` loop.

Can you change the `**kwargs` to be `**_` to indicate that they are not used?

**Comment by kszucs** [comment]
We still have a problem which is definitely solvable but would expand the scope of this PR heavily:


```py
# semantically this should produce two different columns
t.select(x=ibis.random(), y=ibis.random())`
```

```py
# semantically this should produce identical columns
rand = ibis.random()
t.select(x=rand, y=rand)
```

Currently we compile both as `rand() AS x, rand() AS y` which behaves differently on the backends. After this PR the two random values are equal from ibis' perspective, but ideally we should ensure that they produce the same columns on all backends. 
The solution could be to inject an additional subselect:

```sql
SELECT *, randcol AS x, randcol AS y FROM (SELECT *, rand() as randcol FROM ...)  # where randcol name would be generated
```

**Comment by cpcloud** [comment]
Can that be done in a follow up?

**Comment by kszucs** [comment]
> Can that be done in a follow up?

Definitely! That is slightly harder and we need to work around backend specifics.

Yes, please regenerate the snapshots for the failing backends!

**Comment by cpcloud** [comment]
Clouds are passing:

```
…/ibis on  impure is 📦 v8.0.0 via 🐍 v3.11.8 via ❄️   impure (ibis-3.11.8-env)
❯ pytest -m 'bigquery or snowflake' -n auto --dist loadgroup --snapshot-update
==================================================================================== test session starts ====================================================================================
platform linux -- Python 3.11.8, pytest-8.1.1, pluggy-1.4.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Using --randomly-seed=168840051
rootdir: /home/cloud/src/ibis
configfile: pyproject.toml
plugins: xdist-3.5.0, pytest_httpserver-1.0.10, cov-5.0.0, benchmark-4.0.0, randomly-3.15.0, hypothesis-6.100.1, repeat-0.9.3, snapshot-0.9.0, mock-3.14.0, clarity-1.0.1
16 workers [3509 items]
.....................................x......................x....................x.x.....x.....x............x.....x..........xx..................x...xxx............................x [  5%]
.x.........x..x.........................xx.xx.x.xxx...x.xxx.xxx..x....x.....x.....x...x........x.xx.x..x...........x.x......x.......x.x...x..x..x.xx..................x.............. [ 10%]
.x..x...xx....x.x...x....x.............x...................x....x.xx..x........x.x.............x......x..........x...............x.............x......x.............x.x..x........... [ 15%]
....xx.....x........x.xx...x......x..........x........x...........x..x........x...............x.....x...........................xx.......s....x.x..x........................x.x.....x [ 20%]
..x..xx...x..xx.x.xx..x.x..xxx...xx.xxxxx.x.x...xxxxxx....xxxx..xxxx.xxx.x...xxxx..x..x.xx..x...x.x..x..x..xx..xx..x..........x.......x...x......................................x... [ 25%]
.x......x......x...x.....x...xxx.xx.x...xx.x..x.x....x..x..x...........x.........xx.x.......x.....xx............xx..x........x...x......x....x......xxx...x....x.xx...............x.. [ 30%]
.x.......x.........x.............................................s.....s..s......................s...x..s..s...s....x...........................x.......x.........s.................. [ 36%]
................x.........x..x..........x................x..x.......s..................x.........x...................sssssssssssssssssss.sss...x...........x.......x................. [ 41%]
.....................x.....xxx..........x......x....x......x.................x..........x.............................................................x.............................. [ 46%]
........x........x.....x........................................x.x...x.x.xx.x.x..x...xx...x.x...xxx......xxx........x..x..x......x..x...x..............x....x.xx.................... [ 51%]
.........x.....x...........x...x...x.............x...x.x..xx...x........xx.x..x....x.x......x..x.x..x.......x...............x.....x.x..........x.x.....xx..............x............. [ 56%]
.............x...x..................x..x......x...............................x.......x....x.....x.....x.....................x.x...x.......x....x...........s...................xxxx. [ 61%]
.xxx.xx.....xxxx.x.xx.xxxxx....x.xxxx.xxxxxxxxxxxxxxxx.xxxxxxxxxxxx.xxxxxxxxxxxxx.x.xxxxxxxxx.xxxxxxxxxxxxxx.xxxxx....xxxx..x.xx..x.x...x.x.x..x..............x...........x..x.x..... [ 67%]
........x.....x....x.........x.x..xx.xxxx.x.x.x...x..xx..x..x.xx.........x..x....x......x.x..x........x.....x.............x.x.xx.....x...................xx.........x.........x...... [ 72%]
..x.......xxx..........x......xx..............x................s.......................s.................sx..x..sx..x.x.s.sxx.sxxx.x.xx....x.x....x..x.xxx...x....xx..x..x........xx.. [ 77%]
....x..x.s....x........s...s..........s..x.......x...x...........x........x.xxx..........xx.....x........x....x.....x.....x..x....x........x......x.x...x...x........x.x..x..xx...... [ 82%]
......x............x..x...xx..........x......x.......s............x................................................................................x................................. [ 87%]
.....................................................................x..x.xx......................................................................................................... [ 92%]
...........................................................................................................................................................x......................... [ 98%]
..................................................x..................                                                                                                                 [100%]
================================================================= 2921 passed, 45 skipped, 543 xfailed in 356.95s (0:05:56) =================================================================
```

**Review by cpcloud — APPROVED** [review]

**Pull request merged**